### PR TITLE
Feat dragen demux after reset

### PR DESCRIPTION
**Pull request by barrystokman**
## Description
**This PR adds Dragen demultiplexing functionality to cg:**
1. Create a sample sheet suitable for Dragen demultiplexing for a specific flowcell.
2. Create a sample sheet suitable for Dragen demultiplexing for all flowcells in the input directory.
3. Validate a sample sheet suitable for Dragen demultiplexing
4. Start Dragen demultiplexing for a specific flowcell
5. Start Dragen demultiplexing for all eligible flowcells in the input directory

### How to prepare for test
- [x] install this cg branch on hasta-stage: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh feat-dragen-demux`
- [x] install the `cgmodels` branch `feat-dragen-demux` in hasta-stage: 
     1. `git clone -b feat-dragen-demux git@github.com:Clinical-Genomics/cgmodels.git` in `/home/proj/stage/bin/git`
     2. `cg cgmodels`
     3. `poetry build`
     4. `pip install dist/cgmodels-0.7.0.tar.gz`

#### TC1: create dragen samplesheet (dry-run)
- [x] `cg demultiplex samplesheet create --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY --dry-run`

#### TC2: create dragen samplsheet
- [x] `cg demultiplex samplesheet create --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY`

#### TC3: start dragen demultiplexing for a specific flowcell (dry run)
- [x] `cg demultiplex flowcell --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY --dry-run`

#### TC4: start dragen demultiplexing for a specific flowcell
- [x] `cg demultiplex flowcell --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY`

#### TC5: finish dragen demultiplexing for a specific flowcell
preconditions: clear cgstats `cgstats delete -f HNC2FDSXY`
- [x] `cg demultiplex finish flowcell --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY`

#### ~TC6: demultiplex with `BarcodeMismatchIndex1,1`, `BarcodeMismatchIndex2,1`~
- [x] ~`cg demultiplex finish flowcell 210428_A00689_0260_BHNC2FDSXY`~
edit: these settings are default in Dragen BCL Convert)

#### TC7: validate a dragen samplesheet:
- [x]  `cg demultiplex samplesheet validate --bcl-converter dragen /home/proj/stage/flowcells/novaseq/runs/210428_A00689_0260_BHNC2FDSXY/SampleSheet.csv`

#### TC8: create bcl2fastq samplesheet (dry-run) 
- [x] `cg demultiplex samplesheet create 210428_A00689_0260_BHNC2FDSXY --dry-run`

#### TC9: create bcl2fastq samplsheet
- [x] `cg demultiplex samplesheet create 210428_A00689_0260_BHNC2FDSXY`

#### TC10: start bcl2fastq/hasta demultiplexing for a specific flowcell (dry run)
- [x] `cg demultiplex flowcell 210428_A00689_0260_BHNC2FDSXY --dry-run`

#### TC11: start bcl2fastq/hasta demultiplexing for a specific flowcell
- [x] `cg demultiplex flowcell 210428_A00689_0260_BHNC2FDSXY`

#### TC12: finish bcl2fastq/hasta demultiplexing for a specific flowcell
- [x] `cg demultiplex finish flowcell 210428_A00689_0260_BHNC2FDSXY`

#### TC13: validate a bcl2fastq samplesheet:
- [x]  `cg demultiplex samplesheet validate  /home/proj/stage/flowcells/novaseq/runs/210428_A00689_0260_BHNC2FDSXY/SampleSheet.csv`

#### TC1: create dragen samplesheet (dry-run)
- [x] A samplesheet is printed to terminal
- [x] The header for sample id and project are `Sample_ID` and `Sample_Project`

#### TC2: create dragen samplesheet
- [x] `SampleSheet.csv` is created in `/home/proj/stage/flowcells/novaseq/runs/210428_A00689_0260_BHNC2FDSXY`
- [x] The header for sample id and project are `Sample_ID` and `Sample_Project`

#### TC3: start dragen demultiplexing for a specific flowcell (dry run)
- [x] A series of checks are run to see if the flowcell is ready for demultiplexing and the results are printed to the terminal
- [x] the sbatch script is printed to the terminal

#### TC4: start dragen demultiplexing for a specific flowcell
- [x] A series of checks are run to see if the flowcell is ready for demultiplexing and the results are printed to the terminal
- [x]  an sbatch script is created
- [x] an sbatch job is submitted to dragen
- [x] `HNC2FDSXY_demultiplex.stderr` should have no errors

#### TC5: finish dragen demultiplexing for a specific flowcell
- [x] The output directory is restructured `<project_number>` -> `Project_<project_number>`
- [x] Fastq file are placed in sample directories: `Sample<sample_id`>
- [x] The demultiplexing stats are added to `cgstats`
- [x] The demultiplexing stats are written to text files in the output directory
- [x] The demultiplexing stats are similar to the bcl2fast demultiplexing stats in production

#### TC6: demultiplex with `BarcodeMismatchIndex1,1`, `BarcodeMismatchIndex2,1`
- [x] `cg demultiplex finish flowcell 210428_A00689_0260_BHNC2FDSXY`

#### TC7: validate a dragen samplesheet:
- [x]  samplesheet validation for a dragen samplesheet functions correctly

## Review
- [ ] code approved by
- [x] tests executed by @barrystokman 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a- 
- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan
- [x] Document in Atlas: https://clinical-genomics.github.io/atlas/production/data_analysis/workflows/demultiplexing/
- [ ] Deploy this branch
- [ ] Deploy https://github.com/Clinical-Genomics/cgmodels/pull/9
- [ ] Inform to team production

### CODE REVIEW

**Comment by barrystokman** [comment]
Unit tests failing due to undeployed changes in the cgmodels repo :)

**Comment by barrystokman** [comment]
TC1: ✔️  `cg demultiplex samplesheet create --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY --dry-run`
image

TC2: ✔️  ` cg demultiplex samplesheet create --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY`
image
image

TC3: ✔️  `cg demultiplex flowcell --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY --dry-run`
image
image

TC4: ✔️  `cg demultiplex flowcell --bcl-converter dragen 210428_A00689_0260_BHNC2FDSXY`
image
image

**Comment by barrystokman** [comment]
New PR due to reset/merge related issues 👶🏻

This branch is a branch of `feat-dragen-demux` and this PR replaces PR https://github.com/Clinical-Genomics/cg/pull/1204

**Comment by barrystokman** [comment]
demuxing failed after hasta went down: 
image

retest pending paxa reservation.

**Comment by patrikgrenfeldt** [comment]
Great big job Barry! With 62 files I think maybe we go through it together?

**Comment by barrystokman** [comment]
@patrikgrenfeldt Sounds good, but note that this is basically PR https://github.com/Clinical-Genomics/cg/pull/1204 reset to commit https://github.com/Clinical-Genomics/cg/pull/1204/commits/8ee872698b666464a689c304a95e51e96d0dad73. The decision was made to not implement all the feedback in order to put a working version into production and focus on higher prioritized projects.

**Comment by barrystokman** [comment]
All tests pass

File: cg/cli/demultiplex/demux.py [hunk 1]
```diff
@@ -12,16 +14,18 @@
 
 
 @click.command(name="all")
+@OPTION_BCL_CONVERTER
 @click.option("--flowcells-directory", type=click.Path(exists=True, file_okay=False))
 @click.option("--dry-run", is_flag=True)
 @click.pass_obj
 def demultiplex_all(
     context: CGConfig,
+    bcl_converter,
```
Comment (barrystokman) [inline cg/cli/demultiplex/demux.py:23]: type hint

**Comment by barrystokman** [comment]
remark about codefactor issue regarding xml parsing: no untrusted xmls are involved, so not solving it for now.

**Comment by Mropat** [comment]
🔥

**Comment by barrystokman** [comment]
merged, bumped and deployed:
image

**Pull request merged**